doc/compiler/runtime/ELFGenerator.md

```diff
@@ -0,0 +1,183 @@
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# ELFGenerator
+
+This document aims to provide an explanation on how ELF file generation is
+supported in Eclipse OMR. Topics in this include:
+
+1. The two categories of ELF files generated by OMR Compiler
+2. The ELFGenerator class
+3. Sample ELF file generated by OMR Compiler
+
+### Introduction to ELF in OMR Compiler
+
+The ELF (Executable and Linkable Format) standard is the dominating file format
+in Linux, and is also used in many other operating systems. The standard 
+specifies various file formats, such as `.o` object files, executable files,
+`.a` archive files, `.so` shared object files, core dumps, etc. The OMR Compiler
+Technology supports the generating of 2 such variants: executable ELF files, and
+`.o` object files, which we will mainly refer to as relocatable ELF files.
+Executable ELF files can be executed, whereas relocatable ELF files cannot be
+executed unless they are linked into an executable resulting in all the symbolic
+declarations being connected to their symbolic definitions.
+
+The main difference between executable and relocatable ELF files that can be
+generated by OMR Compiler Technology is that executable ELF files have a program
+header that contain information necessary to execute the file, whereas relocatable
+ELF files do not have a program header, but have extra information for relocations.
+Other than that, the 2 variants of ELF files are composed of the same structures.
+
+### ELFGenerator and its Derived Classes
+
+Due to the fact that the 2 variants of ELF files generated by the compiler share
+a lot of similarities, the class ELFGenerator located in the `codegen` component
+has been introduced. The class `ELFGenerator` is an abstract base class that 
+provides a mechanism to share code between different kinds of ELF files.
+Currently the 2 classes that implementation this abstract base class are 
+`ELFExecutableGenerator` and `ELFRelocatableGenerator`. As their names imply, 
+`ELFExecutableGenerator` generates executable ELF, and `ELFRelocatableGenerator`
+generates relocatable ELF.
+
+### How to Enable ELF Generation
+The options `TR_EmitExecutableELFFile` and/or `TR_EmitRelocatableELFFile` will need
+to be enabled if you want the `CodeCacheManager` to write the code cache in memory
+to ELF files. When you want to enable Relocatable ELF file generation, you will also
+need to specify the file name for the `.o` file to write to using the command line
+option `objectFile=<object-file-name.o`. Relocatable ELF Files get generated in the
+working directory, and Executable ELF files are written to
+a file located in `/tmp/`, and named `perf-` followed by the process ID at the time
+of writing the ELF file. The relocatable ELF files are written to 
```

Comment:
For the executable ELF files, perhaps mention that the choice of filename and location is largely historical.  (We should fix this up one day).